### PR TITLE
Read the information from both the definition file and the spec and add logs if have difference

### LINKED ISSUE #18232: Get normalization image from the connector definition file
Get normalization image from the connector definition file and read the information from both the definition file and the spec and add logs if have difference.

### LINKED ISSUE #18235: Read the information from both the definition file and the spec and add logs if have difference.

### PR DESCRIPTION
**Pull request by andriikorotkov**
## What
Added the ability to read the name and tag for the normalization container docker from the definition file, and compare this value with the value obtained from the NormalizationRunnerFactory. In case of difference, the message is added to the logs.

## How
*Describe the solution*

## Recommended reading order
1. `NormalizationRunnerFactory.java`
2. `NormalizationActivityImpl.java`
3. `DefaultNormalizationRunner.java`

## 🚨 User Impact 🚨
No impact

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-commons-temporal/src/main/java/io/airbyte/commons/temporal/TemporalClient.java
Comment (pedroslopez) [inline airbyte-commons-temporal/src/main/java/io/airbyte/commons/temporal/TemporalClient.java:386]: We shouldn't be reading from a definitions provider or direct yamls here- they're mainly used to seed/update the connector definitions in the database (actor_definitions table) and then any other operations within the platform should use the data in the db. For example, cloud doesn't use local definitions and instead loads remote definitions from a service.

I think by this point we should have the normalization tag / repo / anything else in that database table, specially since I see a migration adding those columns to the db... is this true?

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java
Comment (pedroslopez) [inline airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java:55]: for my understanding, this is something that could happen in the meantime before removing NormalizationRunnerFactory because the definitions could have been updated since then, but we would still be using the normalization image from the definitions, right?

Is the NormalizationRunnerFactory or the normalization mappings being used for anything at this point or could we get rid of it as part of this PR?

File: airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml
Comment (pedroslopez) [inline airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml:21]: nit: `supportsDbt`

File: airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml
Comment (pedroslopez) [inline airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml:19]: nit: `normalizationDockerImage`

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java
Comment (pedroslopez) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:87]: Same concern here about using a definitions provider - we should be reading these things from the database

**Review by pedroslopez — CHANGES_REQUESTED** [review]
Thanks andrii! Didn't look too deeply at the logic, but my main concern at the moment is in using local definitions providers instead of reading the data directly from the database. More info in comments!

File: airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml
Comment (andriikorotkov) [reply airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml:21]: :+1:

File: airbyte-commons-temporal/src/main/java/io/airbyte/commons/temporal/TemporalClient.java
Comment (andriikorotkov) [reply airbyte-commons-temporal/src/main/java/io/airbyte/commons/temporal/TemporalClient.java]: Yes you're right. This is a very good point. Thank you for it. I changed this piece of code and now I get the data I need from the database. Also, I removed this logic from **TemporalClient.class** - as far as I understand, that part working on the synchronization process for which we may not retrieve this information.

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java
Comment (andriikorotkov) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java:55]: Yes, we can remove the **NormalizationRunnerFactory** in this pull request. Maybe I'm playing it safe, but I'm doing everything according to the documentation, and it says to do it in separate stages =) And yes, in the future we will use the values from the definition file. Before merging with the master, I will definitely check this point. But as far as I understand, before publishing each new version of the connector, this file is updated, thereby we update the **actor_definition** table. I'm right?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java
Comment (andriikorotkov) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java]: Yep! Thank you. At this point, I changed the value, and in TemporalClient - deleted it. I left the explanation in the comment above.

File: airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/DbtJobOrchestrator.java
Comment (pedroslopez) [inline airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/DbtJobOrchestrator.java:75]: Is normalization version here still the same across the connectors? Specially since we can have different images, I think we would want to have a version per connector.

Comment (pedroslopez) [reply airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/DbtJobOrchestrator.java:75]: From looking at the fields we now have in the DB, I believe this would be coming from `NORMALIZATION_TAG`

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java
Comment (pedroslopez) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:92]: note: there's a utility function `DockerUtils.getTaggedImageName` that deals with combining the repo and tag. Not the most complicated thing to do here but might as well use it if it's available.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java
Comment (pedroslopez) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:89]: Do we have the destinationDefinitionId at this point? If we have that it may be easier / more reliable to get the definition directly by the ID.

**Review by pedroslopez — COMMENTED** [review]
Mainly some questions!

Also, it would be great to see some tests added to make sure that we're retrieving the normalization image/tag as expected from the db

File: airbyte-commons-temporal/build.gradle
Comment (andriikorotkov) [reply airbyte-commons-temporal/build.gradle:18]: Good point, used it for LocalDefinitionsProvider. for this pull request, it is redundant. Thanks for noticing this.

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java
Comment (andriikorotkov) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java:91]: Everything you wrote about is true. I plan to remove the DestinationType enum in the next pull request, where the NormalizationRunnerFactory will also be removed. You must have been misled by the getDestinationTypeByNormalizationImageNamePart method. I will delete it.

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java
Comment (andriikorotkov) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java:59]: Yes, so far I've left using the image from the platform and so far I'm just logging the difference. In the next pull request, **NormalizationRunnerFactory** will be removed along with the create method, and instead, we will directly use the new constructor for `DefaultNormalizationRunner`. (instead of `NormalizationRunnerFactory.create(......)` -> `new DefaultNormalizationRunner(processFactory, normalizationImage)`). and I added the **factoryNormalizationImage** variable to the method so as not to use the String.format(...) construction several times =)

File: airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/DbtJobOrchestrator.java
Comment (andriikorotkov) [reply airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/DbtJobOrchestrator.java:75]: Yes, everything is correct. The new parameter that I added for comparison to the create method already has an image and a tag in it. As I mentioned above, the next step is to remove the create method, and instead, we will directly use the new constructor for DefaultNormalizationRunner. (instead of NormalizationRunnerFactory.create(......) -> new DefaultNormalizationRunner(processFactory, normalizationImage)).

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java
Comment (andriikorotkov) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:92]: I have not yet come across the use of this extra class. Thanks for sharing it)

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java
Comment (andriikorotkov) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:89]: I also tried to find the destination definition id and use the `getStandardDestinationDefinition` method. But in this case, destination definition id doesn't transfer and I can't get it here. Maybe I missed something.

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java
Comment (pedroslopez) [inline airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java:55]: is there any action to take if we get these logs?

File: airbyte-api/src/main/openapi/config.yaml
Comment (pedroslopez) [inline airbyte-api/src/main/openapi/config.yaml:2991]:
```suggestion
        supportsDbt:
          type: boolean
          default: false
```

nit to keep consistent

File: airbyte-api/src/main/openapi/config.yaml
Comment (pedroslopez) [inline airbyte-api/src/main/openapi/config.yaml:3029]:
```suggestion
        supportsDbt:
          type: boolean
          default: false
```

File: airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml
Comment (pedroslopez) [inline airbyte-worker-models/src/main/resources/workers_models/IntegrationLauncherConfig.yaml:23]:
```suggestion
  supportsDbt:
    type: boolean
    default: false
```

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/sync/NormalizationActivityImpl.java
Comment (pedroslopez) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/sync/NormalizationActivityImpl.java:153]: are these error logs intentional?

File: airbyte-api/src/main/openapi/config.yaml
Comment (pedroslopez) [inline airbyte-api/src/main/openapi/config.yaml:2991]: I think this api is called when adding custom destinations in OSS... do we need to add these to the form in the frontend?

What happens if a destination doesn't have these fields set? Are we properly handling this as a destination that doesn't run normalization?

**Review by pedroslopez — APPROVED** [review]
We should standardize to `supportsDbt` (with an `s`) but no need for another review from me for this change.

Please ensure that the tests in CI are passing before merging, seems like the branch is currently red.

Note: after this pr is merged we will still be using the platform-provided normalization image/versions but will start to read the definitions and log if there's a mismatch. A follow-up PR will actually remove the paltform-provided normalization images/versions and just use the ones from the definitions so I'm not expecting this to introduce a user-impacting change just yet.

File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java
Comment (andriikorotkov) [reply airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java:55]: currently - no, just to see that different versions are used. In general, this is necessary for testing and making sure that we are getting the correct data from the actor_definition table

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/sync/NormalizationActivityImpl.java
Comment (andriikorotkov) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/sync/NormalizationActivityImpl.java:153]: yes, I added them on purpose, but I'd probably rather remove them. They are needed more for development than for the end user

**Pull request merged**
